src/mds/SessionMap.h

```diff
@@ -215,20 +216,63 @@ class Session : public RefCountedObject {
     ceph_assert(!info.prealloc_inos.empty());
 
     if (ino) {
-      if (info.prealloc_inos.contains(ino))
+      if (info.prealloc_inos.contains(ino)) {
 	info.prealloc_inos.erase(ino);
-      else
+	if (delegated_inos.contains(ino))
+	  delegated_inos.erase(ino);
+      } else {
 	ino = 0;
+      }
     }
     if (!ino) {
-      ino = info.prealloc_inos.range_start();
-      info.prealloc_inos.erase(ino);
+      /* Grab first prealloc_ino that isn't delegated */
+      for (interval_set<inodeno_t>::const_iterator p = info.prealloc_inos.begin(); p != info.prealloc_inos.end(); ++p) {
```

Comment:
I'd suggest using the C++17 [range-based for loop](https://en.cppreference.com/w/cpp/language/range-for) here.

So, you'd say:
for (const auto p: info.prealloc_inos) {

Reply:
Care to suggest a patch? I tried that and could never get it to work properly here. Changing those lines to read:
```
      for (auto p: info.prealloc_inos) {                                        
        for (auto i = p.first ; i < p.first + p.second ; i += 1) {              
```
...and I get::
```
In file included from /home/jlayton/git/ceph/src/mds/LogSegment.h:19,
                 from /home/jlayton/git/ceph/src/mds/MDLog.h:47,
                 from /home/jlayton/git/ceph/src/mds/Beacon.cc:23:
/home/jlayton/git/ceph/src/include/interval_set.h: In instantiation of ‘std::pair<_FIter, _FIter>& interval_set<T, Map>::iterator::operator*() [with T = inodeno_t; Map = std::map<inodeno_t, inodeno_t, std::less<inodeno_t>, std::allocator<std::pair<const inodeno_t, inodeno_t> > >]’:
/home/jlayton/git/ceph/src/mds/SessionMap.h:224:31:   required from here
/home/jlayton/git/ceph/src/include/interval_set.h:60:25: error: cannot bind non-const lvalue reference of type ‘std::pair<inodeno_t, inodeno_t>&’ to an rvalue of type ‘std::pair<inodeno_t, inodeno_t>’
   60 |                 return *_iter;
      |                         ^~~~~
In file included from /usr/include/c++/9/bits/stl_algobase.h:64,
                 from /usr/include/c++/9/bits/char_traits.h:39,
                 from /usr/include/c++/9/string:40,
                 from /home/jlayton/git/ceph/src/include/ceph_assert.h:5,
                 from /home/jlayton/git/ceph/src/common/dout.h:21,
                 from /home/jlayton/git/ceph/src/mds/Beacon.cc:16:
/usr/include/c++/9/bits/stl_pair.h:291:19: note:   after user-defined conversion: ‘constexpr std::pair<_T1, _T2>::pair(const std::pair<_U1, _U2>&) [with _U1 = const inodeno_t; _U2 = inodeno_t; typename std::enable_if<(std::_PCC<((! std::is_same<_T1, _U1>::value) || (! std::is_same<_T2, _U2>::value)), _T1, _T2>::_ConstructiblePair<_U1, _U2>() && std::_PCC<((! std::is_same<_T1, _U1>::value) || (! std::is_same<_T2, _U2>::value)), _T1, _T2>::_ImplicitlyConvertiblePair<_U1, _U2>()), bool>::type <anonymous> = true; _T1 = inodeno_t; _T2 = inodeno_t]’
  291 |         constexpr pair(const pair<_U1, _U2>& __p)
      |                   ^~~~
make[2]: *** [src/mds/CMakeFiles/mds.dir/build.make:115: src/mds/CMakeFiles/mds.dir/Beacon.cc.o] Error 1
```

Reply:
Fix for that error is here: https://github.com/ceph/ceph/pull/32185

BTW, better is:

````
for (const auto& [first, len] : info.prealloc_inos) {
  for (auto i = first; i < first+len; ++i) {
  }
}
````

Reply:
Thanks for fixing that. I'll plan to rebase this once https://github.com/ceph/ceph/pull/32185 is merged.